# Análise Histórica do Preço do Bitcoin com Aprendizagem de Máquina

Este projeto tem como objetivo analisar o comportamento histórico do preço do Bitcoin utilizando técnicas de análise de dados e visualização gráfica. O dataset utilizado contém informações históricas de negociação do Bitcoin ao longo do tempo.

## Carregamento do Dataset

O dataset utilizado foi obtido da plataforma Kaggle, contendo dados históricos do preço do Bitcoin. As informações incluem preço de abertura, fechamento, máximo, mínimo e volume de negociação.

Os dados foram carregados utilizando a biblioteca Pandas para permitir manipulação e análise eficiente.

In [ ]:
import kagglehub

path = kagglehub.dataset_download("mczielinski/bitcoin-historical-data")

print("Caminho", path)

Using Colab cache for faster access to the 'bitcoin-historical-data' dataset.
Path to dataset files: /kaggle/input/bitcoin-historical-data


In [ ]:
import os

os.listdir(path)

['btcusd_1-min_data.csv']

In [ ]:
import pandas as pd

file_path = os.path.join(path, "btcusd_1-min_data.csv")

df = pd.read_csv(file_path)

df.head()

,Timestamp,Open,High,Low,Close,Volume
0,1.325412e+09,4.58,4.58,4.58,4.58,0.0
1,1.325412e+09,4.58,4.58,4.58,4.58,0.0
2,1.325412e+09,4.58,4.58,4.58,4.58,0.0
3,1.325412e+09,4.58,4.58,4.58,4.58,0.0
4,1.325412e+09,4.58,4.58,4.58,4.58,0.0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7536652 entries, 0 to 7536651
Data columns (total 6 columns):
 #   Column     Dtype  
---  ------     -----  
 0   Timestamp  float64
 1   Open       float64
 2   High       float64
 3   Low        float64
 4   Close      float64
 5   Volume     float64
dtypes: float64(6)
memory usage: 345.0 MB


In [ ]:
df.head()

,Timestamp,Open,High,Low,Close,Volume
0,1.325412e+09,4.58,4.58,4.58,4.58,0.0
1,1.325412e+09,4.58,4.58,4.58,4.58,0.0
2,1.325412e+09,4.58,4.58,4.58,4.58,0.0
3,1.325412e+09,4.58,4.58,4.58,4.58,0.0
4,1.325412e+09,4.58,4.58,4.58,4.58,0.0


In [ ]:
df.describe()

,Timestamp,Open,High,Low,Close,Volume
count,7.536652e+06,7.536652e+06,7.536652e+06,7.536652e+06,7.536652e+06,7.536652e+06
mean,1.551517e+09,2.291424e+04,2.292233e+04,2.290599e+04,2.291426e+04,5.037246e+00
std,1.305474e+08,3.106083e+04,3.106924e+04,3.105236e+04,3.106085e+04,2.176645e+01
min,1.325412e+09,3.800000e+00,3.800000e+00,3.800000e+00,3.800000e+00,0.000000e+00
25%,1.438462e+09,4.527100e+02,4.528600e+02,4.525100e+02,4.527075e+02,2.125472e-02
50%,1.551512e+09,7.791000e+03,7.796850e+03,7.784990e+03,7.790675e+03,4.545712e-01
75%,1.664561e+09,3.657900e+04,3.659800e+04,3.656126e+04,3.657900e+04,2.840000e+00
max,1.777681e+09,1.262020e+05,1.262720e+05,1.261580e+05,1.262020e+05,5.853852e+03


## Conversão de Timestamp

Os registros temporais estavam originalmente no formato Unix Timestamp. Dessa forma, foi realizada a conversão desses valores para o formato de data padrão, permitindo análises temporais adequadas.

In [ ]:
df["Date"] = pd.to_datetime(df["Timestamp"], unit="s")

df.head()

,Timestamp,Open,High,Low,Close,Volume,Date
0,1.325412e+09,4.58,4.58,4.58,4.58,0.0,2012-01-01 10:01:00
1,1.325412e+09,4.58,4.58,4.58,4.58,0.0,2012-01-01 10:02:00
2,1.325412e+09,4.58,4.58,4.58,4.58,0.0,2012-01-01 10:03:00
3,1.325412e+09,4.58,4.58,4.58,4.58,0.0,2012-01-01 10:04:00
4,1.325412e+09,4.58,4.58,4.58,4.58,0.0,2012-01-01 10:05:00


## Limpeza dos Dados

Nesta etapa, foram removidos registros contendo valores ausentes. Essa etapa é fundamental para garantir a qualidade e confiabilidade das análises realizadas posteriormente.

In [ ]:
df = df.dropna()
df.isnull().sum()

,0
Timestamp,0
Open,0
High,0
Low,0
Close,0
Volume,0
Date,0


In [ ]:
print(df.columns.tolist())

['Timestamp', 'Open', 'High', 'Low', 'Close', 'Volume', 'Date']


## Agregação Temporal dos Dados

Os dados originais estavam registrados em intervalos de minutos, o que resultava em um grande volume de informações. Dessa forma, foi realizada uma agregação temporal para transformar os dados em registros diários.

Essa transformação reduz a complexidade computacional e facilita a visualização e análise do comportamento do preço ao longo do tempo.

In [ ]:
df_daily = df.resample("D", on="Date").agg({
    "Open": "first",
    "High": "max",
    "Low": "min",
    "Close": "last",
    "Volume": "sum"
}).dropna()

df_daily = df_daily.reset_index()

df_daily.head()

,Date,Open,High,Low,Close,Volume
0,2012-01-01,4.58,4.84,4.58,4.84,10.000000
1,2012-01-02,4.84,5.00,4.84,5.00,10.100000
2,2012-01-03,5.00,5.32,5.00,5.29,107.085281
3,2012-01-04,5.29,5.57,4.93,5.57,107.233260
4,2012-01-05,5.57,6.46,5.57,6.42,70.328742


## Visualização do Preço do Bitcoin

Após o pré-processamento dos dados, foi gerado um gráfico representando a evolução histórica do preço de fechamento do Bitcoin.

A biblioteca Plotly foi utilizada para gerar visualizações interativas, permitindo melhor compreensão das tendências ao longo do tempo.

In [ ]:
import plotly.express as px

fig = px.line(
    df_daily,
    x="Date",
    y="Close",
    title="Preço diário de fechamento do Bitcoin"
)

fig.show()

## Dimensão do Dataset após Agregação

Após a agregação temporal dos dados, o dataset passou a conter 5230 registros diários e 10 atributos. Cada linha representa um dia de negociação do Bitcoin, contendo informações como abertura, máxima, mínima, fechamento e volume.

Essa transformação permitiu reduzir o volume original dos dados, que estavam em frequência de minutos, tornando a análise mais eficiente e adequada para visualizações e modelos de aprendizagem de máquina.

In [ ]:
df_daily.shape

(5236, 6)

## Cálculo dos Retornos Diários

Além do preço de fechamento, também foram calculados os retornos diários do Bitcoin. O retorno representa a variação percentual do preço entre dois dias consecutivos.

Também foi calculado o retorno logarítmico, frequentemente utilizado em análises de séries temporais financeiras por facilitar a interpretação das variações relativas ao longo do tempo.

In [ ]:
import numpy as np

df_daily["Return"] = df_daily["Close"].pct_change()
df_daily["Log_Return"] = np.log(df_daily["Close"] / df_daily["Close"].shift(1))

df_daily = df_daily.dropna()

df_daily.head()

,Date,Open,High,Low,Close,Volume,Return,Log_Return
1,2012-01-02,4.84,5.00,4.84,5.00,10.100000,0.033058,0.032523
2,2012-01-03,5.00,5.32,5.00,5.29,107.085281,0.058000,0.056380
3,2012-01-04,5.29,5.57,4.93,5.57,107.233260,0.052930,0.051577
4,2012-01-05,5.57,6.46,5.57,6.42,70.328742,0.152603,0.142023
5,2012-01-06,6.42,6.90,6.40,6.40,55.855834,-0.003115,-0.003120


## Visualização dos Retornos Logarítmicos

Após o cálculo dos retornos logarítmicos, foi realizada a visualização da variação diária do Bitcoin ao longo do tempo.

Esse tipo de visualização permite identificar períodos de alta volatilidade, nos quais ocorrem variações mais intensas no preço do ativo.

In [ ]:
import plotly.express as px

fig = px.line(
    df_daily,
    x="Date",
    y="Log_Return",
    title="Retorno Logarítmico Diário do Bitcoin"
)

fig.show()

## Distribuição dos Retornos

Foi gerado um histograma dos retornos logarítmicos com o objetivo de observar a distribuição das variações diárias do preço do Bitcoin.

Essa análise permite verificar se os retornos apresentam concentração próxima de zero ou se existem valores extremos associados a períodos de alta volatilidade.

In [ ]:
fig = px.histogram(
    df_daily,
    x="Log_Return",
    nbins=100,
    title="Distribuição dos Retornos Logarítmicos"
)

fig.show()

## Cálculo das Médias Móveis

As médias móveis são utilizadas para suavizar variações de curto prazo e destacar tendências ao longo do tempo.

Neste trabalho, foram calculadas médias móveis de 7 e 30 dias, representando tendências de curto e médio prazo.

In [ ]:
df_daily["MA_7"] = df_daily["Close"].rolling(window=7).mean()
df_daily["MA_30"] = df_daily["Close"].rolling(window=30).mean()

fig = px.line(
    df_daily,
    x="Date",
    y=["Close", "MA_7", "MA_30"],
    title="Preço do Bitcoin com Médias Móveis"
)

fig.show()

## Análise de Autocorrelação

A autocorrelação foi calculada com o objetivo de verificar se os retornos do Bitcoin apresentam dependência temporal.

Essa análise permite identificar se valores passados influenciam valores futuros, sendo um aspecto importante na verificação do comportamento de caminhada aleatória (random walk).

In [ ]:
import plotly.express as px
import pandas as pd

lag_max = 30

acf_values = [
    df_daily["Log_Return"].autocorr(lag=i)
    for i in range(lag_max)
]

acf_df = pd.DataFrame({
    "Lag": range(lag_max),
    "Autocorrelation": acf_values
})

fig = px.bar(
    acf_df,
    x="Lag",
    y="Autocorrelation",
    title="Autocorrelação dos Retornos Logarítmicos"
)

fig.show()

## Cálculo do Expoente de Hurst

O expoente de Hurst foi calculado com o objetivo de avaliar a presença de dependência de longo prazo na série temporal.

Esse indicador permite identificar se a série apresenta comportamento aleatório, tendência persistente ou reversão à média, sendo amplamente utilizado em análises financeiras e estudos de séries temporais.

In [ ]:
!pip install hurst

O cálculo inicial do Expoente de Hurst utilizando o preço de fechamento bruto resultou em valor acima do intervalo esperado. Por esse motivo, a análise foi ajustada para utilizar os retornos logarítmicos, que são mais adequados para séries financeiras por representarem variações relativas entre períodos consecutivos.

In [ ]:
from hurst import compute_Hc

serie = df_daily["Log_Return"].dropna().values

H, c, data = compute_Hc(
    serie,
    kind="change",
    simplified=True
)

print("Expoente de Hurst:", H)

Expoente de Hurst: 0.6479043463619991


## Cálculo do Expoente de Hurst

O expoente de Hurst foi calculado utilizando os retornos logarítmicos do preço de fechamento do Bitcoin, com o objetivo de avaliar a presença de dependência de longo prazo na série temporal.

O valor obtido para o expoente de Hurst foi aproximadamente H = 0.648. Esse resultado indica comportamento persistente, sugerindo que a série apresenta dependência temporal de longo prazo.

Esse comportamento indica que variações passadas podem influenciar tendências futuras, embora não de forma determinística, sendo característico de séries financeiras reais.

## Preparação dos Dados para Machine Learning

Para realizar a previsão do preço do Bitcoin, foram criadas variáveis baseadas em valores passados da série temporal. Essas variáveis, conhecidas como lags, representam valores anteriores que podem ser utilizados como entrada para modelos de aprendizagem de máquina.

Essa abordagem permite que o modelo identifique padrões temporais presentes nos dados históricos.

In [ ]:
df_ml = df_daily.copy()

df_ml["Close_lag1"] = df_ml["Close"].shift(1)
df_ml["Close_lag2"] = df_ml["Close"].shift(2)
df_ml["Close_lag3"] = df_ml["Close"].shift(3)

df_ml = df_ml.dropna()

df_ml.head()

,Date,Open,High,Low,Close,Volume,Return,Log_Return,MA_7,MA_30,Close_lag1,Close_lag2,Close_lag3
30,2012-01-31,5.58,6.22,5.55,5.55,104.718333,-0.005376,-0.005391,5.727143,6.364000,5.58,5.90,4.91
31,2012-02-01,5.55,6.00,3.80,5.99,13.552446,0.079279,0.076293,5.725714,6.397000,5.55,5.58,5.90
32,2012-02-02,5.99,6.26,5.54,6.26,123.582638,0.045075,0.044089,5.724286,6.429333,5.99,5.55,5.58
33,2012-02-03,6.26,6.35,5.93,6.29,283.382106,0.004792,0.004781,5.782857,6.453333,6.26,5.99,5.55
34,2012-02-04,6.29,6.50,5.94,6.50,67.694994,0.033386,0.032841,6.010000,6.456000,6.29,6.26,5.99


In [ ]:
df_ml.shape

(5206, 13)

## Divisão entre Treino e Teste

Como se trata de uma série temporal, os dados foram divididos mantendo a ordem cronológica. Os primeiros 80% dos registros foram utilizados para treinamento do modelo, enquanto os 20% finais foram usados para teste.

Essa abordagem evita o embaralhamento dos dados e simula melhor um cenário real de previsão.

In [ ]:
X = df_ml[["Close_lag1", "Close_lag2", "Close_lag3"]]
y = df_ml["Close"]

split = int(len(df_ml) * 0.8)

X_train = X.iloc[:split]
X_test = X.iloc[split:]

y_train = y.iloc[:split]
y_test = y.iloc[split:]

print("Treino:", X_train.shape)
print("Teste:", X_test.shape)

Treino: (4164, 3)
Teste: (1042, 3)


## Treinamento do Modelo Random Forest

Foi utilizado o algoritmo Random Forest Regressor para prever o preço de fechamento do Bitcoin. Esse modelo combina várias árvores de decisão para gerar uma previsão mais robusta e reduzir o risco de sobreajuste em relação a uma única árvore.

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error

model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred) ** 0.5

print("MAE:", mae)
print("RMSE:", rmse)

MAE: 16286.175959980852
RMSE: 24215.869278636445


In [ ]:
df_daily["Close"].mean()

np.float64(22941.96946131805)

## Ajuste da Variável Alvo

Inicialmente, o modelo foi treinado para prever diretamente o preço de fechamento do Bitcoin. No entanto, séries financeiras apresentam crescimento ao longo do tempo, o que pode dificultar a aprendizagem do modelo.

Dessa forma, foi adotada uma abordagem alternativa, utilizando os retornos logarítmicos como variável alvo, permitindo uma modelagem mais adequada das variações relativas da série temporal.

In [ ]:
df_ml2 = df_daily.copy()

df_ml2["Return_lag1"] = df_ml2["Log_Return"].shift(1)
df_ml2["Return_lag2"] = df_ml2["Log_Return"].shift(2)
df_ml2["Return_lag3"] = df_ml2["Log_Return"].shift(3)

df_ml2 = df_ml2.dropna()

In [ ]:
X = df_ml2[
    [
        "Return_lag1",
        "Return_lag2",
        "Return_lag3"
    ]
]

y = df_ml2["Log_Return"]

In [ ]:
split = int(len(df_ml2) * 0.8)

X_train = X.iloc[:split]
X_test = X.iloc[split:]

y_train = y.iloc[:split]
y_test = y.iloc[split:]

## Importância das Variáveis

Foi realizada a análise de importância das variáveis utilizadas no modelo Random Forest, com o objetivo de identificar quais valores passados exercem maior influência na previsão dos retornos futuros.

Essa análise permite compreender melhor o comportamento temporal da série.

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error

model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred) ** 0.5

print("MAE:", mae)
print("RMSE:", rmse)

MAE: 0.019921161668881342
RMSE: 0.027227142025811942


## Avaliação do Modelo com Retornos Logarítmicos

Após o ajuste da variável alvo para retornos logarítmicos, o modelo Random Forest apresentou desempenho significativamente melhor.

Os resultados obtidos foram:

- MAE ≈ 0.01987
- RMSE ≈ 0.02718

Esses valores indicam que o modelo consegue prever variações relativas com erro médio aproximado de 2%, demonstrando capacidade moderada de capturar padrões presentes na série temporal.

In [ ]:
import pandas as pd
import plotly.express as px

importance = pd.DataFrame({
    "Feature": X.columns,
    "Importance": model.feature_importances_
})

importance = importance.sort_values(
    by="Importance",
    ascending=False
)

fig = px.bar(
    importance,
    x="Feature",
    y="Importance",
    title="Importância das Variáveis no Modelo"
)

fig.show()

importance

,Feature,Importance
0,Return_lag1,0.388998
2,Return_lag3,0.306718
1,Return_lag2,0.304284


## Importância das Variáveis

A Figura X apresenta a importância relativa das variáveis utilizadas no modelo Random Forest.

Observa-se que o retorno do dia anterior (Return_lag1) apresentou maior relevância na previsão dos retornos futuros, seguido pelos retornos dos dois e três dias anteriores.

Esse comportamento indica a presença de dependência temporal de curto prazo, sugerindo que eventos recentes possuem maior influência na dinâmica da série temporal.

## Comparação entre Retornos Reais e Previstos

Foi realizada a comparação entre os retornos reais e os retornos previstos pelo modelo Random Forest.

Essa visualização permite avaliar a capacidade do modelo em acompanhar as variações da série temporal.

In [ ]:
results = pd.DataFrame({
    "Date": df_ml2["Date"].iloc[split:],
    "Real": y_test,
    "Predito": y_pred
})

fig = px.line(
    results,
    x="Date",
    y=["Real", "Predito"],
    title="Retornos Reais vs Retornos Previstos"
)

fig.show()

## Comparação entre Retornos Reais e Previstos

A Figura X apresenta a comparação entre os retornos reais e os retornos previstos pelo modelo Random Forest.

Observa-se que o modelo foi capaz de acompanhar o comportamento geral da série temporal, demonstrando capacidade de capturar padrões básicos presentes nos dados históricos.

No entanto, o modelo apresentou maior dificuldade na previsão de eventos extremos, especialmente em períodos de alta volatilidade. Esse comportamento é esperado em séries financeiras, nas quais variações abruptas são frequentemente associadas a fatores externos e imprevisíveis.

## Síntese dos Resultados Obtidos

Os resultados obtidos indicam que o preço do Bitcoin apresenta características típicas de séries financeiras, incluindo alta volatilidade e presença de dependência temporal de curto prazo.

A análise de autocorrelação indicou baixa dependência temporal imediata, enquanto o cálculo do expoente de Hurst revelou comportamento persistente de longo prazo (H ≈ 0.648).

O modelo Random Forest foi capaz de prever retornos logarítmicos com erro médio aproximado de 2%, demonstrando capacidade moderada de capturar padrões presentes na série temporal. Entretanto, o modelo apresentou maior dificuldade na previsão de eventos extremos.

In [ ]:
import os

os.listdir(path)

['btcusd_1-min_data.csv']